doc(TUI-252): add service proposal (#1996)
Talend/ui: packages/cmf/proposal/Service-TUI-252.md

# Build *Service* to organize code by feature

Original Jira: https://jira.talendforge.org/browse/TUI-252

Objective: Provide tools to organize code by feature.
Constraints: must work in react-redux, thunk and redux-saga environment.

## Why do we need service

> When a frontend need to implement a feature he needs to split the needs in two parts:

Reviewer: ```suggestion
When a frontend developer needs to implement a feature he needs to split the needs in two parts:
```

* how do I display it (aka components)
* how do I handle the business (aka services)

The word `service` is very often used to point out two things:

* the state, or the business data
* the api exposed to handle the business objects.

## How should we see service in our environment

> A feature mean the following technical *items* to be defined:

Reviewer: ```suggestion
A feature means the following technical *items* need to be defined:
```

* **initial state** to store your business objects
* action **TYPES** constant to handle changes in that data structure

> * **reducer** which recieve the action and do the write in the redux state

Reviewer: `receives`, `modifies`/`updates`

> * **action creators** to trigger those changes. They will be dispatched by component event handler or internal life cycle

Reviewer: Agree with @jsomsanith better to avoid confusion here (because I've seen it in projects)

* **selectors** to be able to read the data and access it without having to know where it is

> * **saga** are started with the app to handle effects

Reviewer: Furthermore, below is mentionned "_a_ saga" so the concept of "*main (service) saga*" could better show the _modularity_ principle of services.

Theses *items* need boiler plate the developer has to perfectly know the redux echo system.

> Action *TYPE*, *Action creators*, *selectors* are considered as external API. They must be exposed to be used.

Reviewer: ```suggestion
*Action TYPES*, *Action creators*, *selectors* are considered as external API. They must be exposed to be used.
```

The reducer and the saga, must be passed to cmf.bootstrap for the setup.

> So the main idea here is so to provide higher level API which generate the boiler plate for you.

Reviewer: ```suggestion
So the main idea here is so to provide higher level API which generate the boilerplate code for the developer.
```

> The user must be able to write at anytime any custom *items*.

Reviewer: ```suggestion
The developer must be able to write at anytime any custom *items*.
```

## How to write a service

```javascript
export const cmfModule = {

>     id: 'userService', // must be uniq accross the app

Reviewer: ```suggestion
    id: 'userService', // must be unique accross the app
```

    saga: function* mainSaga() {},
    reducer: {
        // TODO: should we force only one path in redux store ?
        // If not, generator must take the attr name as param.
        // so here the path in the store will be state.user
        user: function (state, action) {
            // reduce the action
            // it will be composed with generated reducer
        },
};
const service = {
    // you can write here your custom stuff
    actionCreators: {
        doFetch: function() {},
    },
    selectors: {
        getItems: function(state) {},
    },
};
// call a generator to init boilerplate

> cmf.service.generators.simpleItem(service, cmfModule);

Reviewer 2: I think we should split this into 3 parts instead of 1 exemple with everything : 
- how to write a service : cmfModule + service
- how can I register my action creators in cmf
- how can I avoid some boilerplates with generators

cmf.service.registerService(service, cmfModule);
// ==
cmfModule.actionCreators = {
    'userService.doFetch': service.actionCreators.doFetch,
    'userService.get': service.actionCreators.get,
    'userService.patch': service.actionCreators.get,
    'userService.set': service.actionCreators.get,
};

export default service;
```

## How to use a service

```javascript
import React from 'react';
import { connect } from 'react-redux';
import datasetService from '@talend/dataset/lib/service';

class MyComponent extends React.Component {
    constructor(props) {
        super(props);
        this.state = {}
    }

    componentDidMount() {
        this.props.getDataset();
    }

    render() {
        return (
            <ul>
                {this.props.items.map(item => (
                    <li key={item.id}>{item.name}</li>
                ))}
            </ul>
        );
    }
}

function mapStateToProps(state) {
    return {
        items: datasetService.selectors.get(state),
    };
}

const mapDispatchToProps = {
    getDataset: datasetService.actionCreators.get,
};

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(MyComponent);
```

> ## Usefull Features / needed generators

Reviewer: ```suggestion
## Useful features / needed generators
```
